OOT Link Import/Export + Flipbook Import/Export + Animation Updates/Fixes

kurethedead opened the pull request:
This PR allows for the importing of missing face/mouth textures, which were cases of textures animated through code. Provided an overlay, Fast64 will try to find texture arrays used for flipbook animation and read them into the material. The material will allow you to preview any texture in the flipbook. This will also export correctly, and try to override texture arrays in the provided overlay. Flipbook textures using color index will automatically use a shared palette.

Additionally, OOT skeleton and DL imports will now automatically search in assets/objects/... files that have been included in the object/actor file. Ideally the user should now be able to import skeleton -> modify -> export -> compile without having to hand edit code.
<img width="320" alt="blender_wTff26AqJ5" src="https://user-images.githubusercontent.com/59840896/182973286-bca16bda-7a0b-4328-8c1e-a7a2bd730c04.png">

kurethedead commented:
I've also now added support for Link animation import/export, which includes texture animations that use the flipbook system.

kurethedead commented:
I've added some fixes for some skeleton related bugs, including bad normals on export, incorrect animations on non rotation applied armatures, and negative hex values.

krm01 commented:
I get this error when trying to export:
![image](https://user-images.githubusercontent.com/8379626/188305114-cf3221e3-0944-4227-9e1f-19c8215f2246.png)

krm01 commented on fast64_internal/oot/oot_model_classes.py:
```diff
@@ -1,191 +1,531 @@
+    def onMaterialCommandsBuilt(self, fMaterial, material, drawLayer):
+        # handle dynamic material calls
+        gfxList = fMaterial.material
+        matDrawLayer = getattr(material.flipbookGroup, drawLayer.lower())
```
I think this line was just a typo, I changed this line back to what it was originally and it seems to have fixed the error I was getting.
```suggestion
        matDrawLayer = getattr(material.ootMaterial, drawLayer.lower())
```

kurethedead replied:
Whoops, you're right. Somehow slipped through the cracks.

krm01 commented on fast64_internal/oot/oot_model_classes.py:
```diff
@@ -1,191 +1,531 @@
+
+    def processFlipbookCI(self, fMaterial: FMaterial, flipbookProp: FlipbookProperty, texProp: TextureProperty):
+        # print("Processing flipbook...")
+        flipbook = OOTTextureFlipbook(flipbookProp.name, flipbookProp.exportMode, [])
```
typo!
```suggestion
        flipbook = TextureFlipbook(flipbookProp.name, flipbookProp.exportMode, [])
```

kurethedead replied:
I swear, I must have messed up somewhere because I remember making these changes previously.

krm01 commented:
I am not really equipped to give a code review, but I have been testing this branch out pretty extensively over the past several days and as of the latest commit [1d6d659](https://github.com/Fast-64/fast64/pull/126/commits/1d6d659a8a72ceaa53e4a7f73cf7ddb249101e1f) all the features in this PR are working correctly. Importing is fine, replacing eye/mouth textures with brand new ones, assigning a brand new model to the skeleton and exporting all works!

Dragorn421 reviewed (commented):
A few comments

#136 will probably get merged first, and bring conflicts about export/import settings. Should be okay, right?

kurethedead commented on fast64_internal/utility_anim.py:
```diff
@@ -7,6 +8,91 @@
+    return (restRotation.inverted() @ inputEuler.to_matrix().to_4x4()).to_euler("XYZ", inputEuler)
+
+
+def attemptModifierApply(modifier):
```
I think I needed it for `armatureApplyWithMesh`, which resulted in a circular dependency between `utility` and `utility_anim` unless I moved it.

kurethedead commented:
I'm fine with #136 being merged first, I'll handle the conflicts.

kurethedead commented:
Since this PR is the first time fast64 has Link import/export support, I wanted to add some commits from other branches that are minor but would be technically breaking changes. Skeleton and bone properties are moved into their own property groups, and actor scale is now an object property instead of a scene property. I've also added the ability to import Link in an A-pose for convenience.

Dragorn421 commented on fast64_internal/oot/oot_anim.py:
```diff
@@ -494,30 +903,20 @@ def oot_anim_panel_unregister():
-    del bpy.types.Scene.ootAnimIsCustomExport
-    del bpy.types.Scene.ootAnimExportCustomPath
-    del bpy.types.Scene.ootAnimExportFolderName
+    bpy.types.Object.ootLinkTextureAnim = bpy.props.PointerProperty(type=OOTLinkTextureAnimProperty)
```
this can go into `OOT_ObjectProperties` (access with `object.fast64.oot`)

Dragorn421 commented on fast64_internal/oot/oot_f3d_writer.py:
```diff
@@ -19,6 +21,9 @@ class OOTDLExportSettings(bpy.types.PropertyGroup):
     removeVanillaData: bpy.props.BoolProperty(name="Replace Vanilla DLs")
     drawLayer: bpy.props.EnumProperty(name="Draw Layer", items=ootEnumDrawLayers)
+    overlay: bpy.props.StringProperty(name="Overlay", default="")
+    is2DArray: bpy.props.BoolProperty(name="Has 2D Flipbook Array", default=False)
```
```suggestion
    flipbookUses2DArray: bpy.props.BoolProperty(name="Has 2D Flipbook Array", default=False)
```

Dragorn421 commented on fast64_internal/oot/oot_f3d_writer.py:
```diff
@@ -19,6 +21,9 @@ class OOTDLExportSettings(bpy.types.PropertyGroup):
     drawLayer: bpy.props.EnumProperty(name="Draw Layer", items=ootEnumDrawLayers)
+    overlay: bpy.props.StringProperty(name="Overlay", default="")
+    is2DArray: bpy.props.BoolProperty(name="Has 2D Flipbook Array", default=False)
+    arrayIndex2D: bpy.props.IntProperty(name="Index if 2D Array", default=0, min=0)
```
```suggestion
    flipbookArrayIndex2D: bpy.props.IntProperty(name="Index if 2D Array", default=0, min=0)
```

Dragorn421 commented on fast64_internal/oot/oot_f3d_writer.py:
```diff
@@ -19,6 +21,9 @@ class OOTDLExportSettings(bpy.types.PropertyGroup):
     isCustom: bpy.props.BoolProperty(name="Use Custom Path")
     removeVanillaData: bpy.props.BoolProperty(name="Replace Vanilla DLs")
     drawLayer: bpy.props.EnumProperty(name="Draw Layer", items=ootEnumDrawLayers)
+    overlay: bpy.props.StringProperty(name="Overlay", default="")
```
```suggestion
    actorOverlayName: bpy.props.StringProperty(name="Actor Overlay", description="Name of the actor overlay, for importing eye/mouth textures as flipbooks. For example ovl_En_Ta" default="")
```

Dragorn421 commented on fast64_internal/oot/oot_texture_array.py:
```diff
@@ -0,0 +1,235 @@
+
+# Special cases:
+# z_en_xc: one texture is not stored in any array.
+# skeletonName only used for en_ossan (shopkeepers) and demo_ec (end credits party), which have multiple actors in it
```
```suggestion
# skeletonName only used for en_ossan (shopkeepers) and demo_ec (end credits party), which have multiple skeletons
```

kurethedead commented on fast64_internal/f3d/flipbook.py:
```diff
@@ -0,0 +1,329 @@
+def flipbook_2d_to_c(flipbook: TextureFlipbook, isStatic: bool, count: int):
+    newArrayData = "void* " if not isStatic else "static void* "
+    newArrayData += f"{flipbook.name}[][{len(flipbook.textureNames)}] = {{ "
+    for i in range(count):
+        newArrayData += "{ " + flipbook_data_to_c(flipbook) + " },\n"
```
This is kind of a weird situation; modifying a 2D array makes sense, but with a custom export the array is built from scratch (this function). Having just one element might be confusing to the user if they're copy pasting code over, which is why I just repeated the element up until its index. I'm not sure what the best option is.

Dragorn421 replied:
Did you notice the `* count` I added?

kurethedead replied:
Oh nvm, my bad lmao

kurethedead commented on fast64_internal/f3d/flipbook.py:
```diff
@@ -0,0 +1,329 @@
+
+
+def drawTextureArray(
+    layout: bpy.types.UILayout, textureArray: bpy.types.CollectionProperty, index: int, exportMode: str
```
That type should still be technically correct, I think? If not descriptive? I feel like it would be better to leave it as opposed to not having a type at all.

Dragorn421 replied:
yeah I agree it's a shame there's no good choice to type it, one option would be to add our own type, maybe in utility.py, like `CollectionProperty = Any  # collection prop as defined by using bpy.props.CollectionProperty` maybe

kurethedead commented on fast64_internal/f3d/flipbook.py:
```diff
@@ -0,0 +1,329 @@
+    textures: bpy.props.CollectionProperty(type=ImagePointerProperty)
+
+
+# The reason these are separate is for the case when the user changes the material draw layer, but not the
+# dynamic material calls. This could cause crashes which would be hard to detect.
```
Honestly I'm not sure what I was saying here 🤔I think I was talking about why the flipbook property group was not part of the OOT dynamic material property group, but I eventually moved flipbook stuff to be game agnostic anyway so the message is outdated.

kurethedead commented on fast64_internal/f3d/flipbook.py:
```diff
@@ -0,0 +1,329 @@
+        register_class(cls)
+
+    bpy.app.handlers.frame_change_pre.append(flipbookAnimHandler)
+    bpy.types.Material.flipbookGroup = bpy.props.PointerProperty(type=FlipbookGroupProperty)
```
I was thinking that too, but its not really related to F3D at all so I think its better to have it separate.

Dragorn421 replied:
we could have a `material.fast64` `Fast64_MaterialProperties` though, maybe?

Like
- `scene.fast64` `Fast64_Properties`
- `bone.fast64` `Fast64_BoneProperties`
- `object.fast64` `Fast64_ObjectProperties`

kurethedead replied:
Honestly I'm kinda iffy on the bone/object versions in general. Most of those properties are already outside the fast64 property groups, and can't be changed easily since people are already using them (unlike scene, which is mostly operator settings). If some properties are in the fast64 property group and some aren't, then it defeats the purpose of having a fast64 property group in the first place.

Dragorn421 reviewed (commented):
You know, this PR should primarily be named "Link import/export"

kurethedead commented on fast64_internal/f3d/f3d_parser.py:
```diff
@@ -517,6 +525,25 @@ def __init__(self, f3d, basePath, materialContext):
+        self.limbToBoneName = {}
+
+        self.verts = []
+        self.limbGroups = {}
```
They're reset in the clearMaterial() call, but yeah, I think your idea works better.

kurethedead commented:
I've added typing to F3DContext in the process of modifying clearGeometry(), and ended up refactoring a lot of tuple usage into actual classes.

kurethedead commented:
I've added some of the import optimizations from the scene import branch and fixed an issue with imports where the dynamic material calls weren't being handled correctly. I promise I won't add any more to this branch unless its a bug fix or documentation.

kurethedead commented:
I've fixed the current merge conflicts with the new commits.

Dragorn421 commented on fast64_internal/oot/oot_anim.py:
```diff
@@ -213,7 +294,70 @@ def ootConvertAnimationData(anim, armatureObj, convertTransformMatrix, *, frame_
+
+    bpy.context.scene.frame_set(currentFrame)
+
+    # need to deepcopy?
```
deepcopy what?

Zelllll reviewed (approved):
Approving on the basis that this functionally works.

kurethedead commented on fast64_internal/f3d/f3d_material.py:
```diff
@@ -3252,6 +3373,57 @@ class F3DMaterialProperty(bpy.types.PropertyGroup):
+            self.set_key,
+            self.set_k0_5,
+            self.set_combiner,
+            self.set_k0_5,
```
I added a commit to fix the other things, but `use_global_fog` is a weird property in that it's sm64 specific and not technically part of the material itself.

kurethedead commented on fast64_internal/f3d/f3d_parser.py:
```diff
@@ -690,16 +730,20 @@ def addTriangle(self, indices, dlData):
 
-        for material in self.materials:
-            if propertyGroupEquals(self.materialContext.f3d_mat, material.f3d_mat):
+        key = self.getMaterialKey(self.materialContext)
+        if key in self.materialDict:
+            material = self.materialDict[key]
```
`materialContext` is just a regular blender material. Whenever a new material needs to be made, a copy is created and stored in a list. If a matching is key is detected, then the existing material in that list is used instead of creating a new copy.

Dragorn421 commented:
Unexpectedly delaying this a bit to slightly improve a `gEmptyList` vs `NULL` behavior
(feel free to correct me Kure and if you want to do it later we can also merge this as is)

krm01 commented:
@Dragorn421 my understanding is the `gEmptyList` vs `NULL` is only a problem for people who tried to use this branch before Kure made the change, and had existing blend files with the old behavior. That's what happened to me. I have been testing this again tonight and a clean import of the skeleton works fine.

kurethedead commented:
Yeah we can do it later, for now I'll just add a warning message.

ariahiro64 commented:
I can confirm that the issue I was having was due to an older skeleton. The skeleton itself was updated successfully so I guess only a small warning is needed and it's good to go.
